GGRC-6361 Add optional description field for evidence URL/File

natalya-mihalyonok opened the pull request:
<!-- If your PR depends on other tickets or PRs, you can list them here
# Dependencies

This PR is `on hold` until the dependencies are merged:

- [ ] GGRC-1234
- [ ] #1234
-->

# Issue description

What   is the problem / issue?   
Although we can use the Response/Comment field to add comments, I believe it   would be easier to have a dedicated field for small comments (ex: "See   section x.y" or "Go to page x")       
What is the impact if it is not fixed / implemented? 
(Quantify the impact,   eg, # of hours)   Minor confusion could occur when adding multiple links/files.

Solution to be implemented:
AC1: Add on Evidence file and evidence url INFO PAGE / INFO PANE/ EDIT EVIDENCE POPUP additional field named 'Notes'. Field type 'Text'
AC2: Add on Assessments/Audits INFO PANE / INFO PAGE / EDIT POPUPs near added Evidence URL/file an 'Edit' icon.
AC3: On 'Edit' icon click a popover with text field, Approve and Cancel buttons should appear.
AC4: After adding a text to popover and clicking 'Approve' a text should appear under 'Evidence file/URL'.
AC5:  After adding a text to popover and clicking 'Cancel' a text should NOT appear under 'Evidence file/URL'.

# Steps to test the changes

1. Open any Assessment/Audit.
2. Click Add Evidence Url/Attach Evidence File.
3. After adding/attaching click 'Edit' icon near the url/file.
4. Verify that a popover with text field, Approve and Cancel buttons are shown. |
5. Enter any notes and click 'Approve'. 
6. Verify that notes appears under url/file. 
7. Verify that notes fields is shown on the Evidence Info. 

# Solution description
FE:
1. Added on Evidence file and evidence url INFO PAGE / INFO PANE/ EDIT EVIDENCE POPUP additional field named 'Notes'. Field type 'Text'
2. Added on Assessments/Audits INFO PANE / INFO PAGE / EDIT POPUPs near added Evidence URL/file an 'Edit' icon.
3. On 'Edit' icon click a popover with text field, Approve and Cancel buttons should appear.
4. After adding a text to popover and clicking 'Approve' a text should appear under 'Evidence file/URL'.
5. After adding a text to popover and clicking 'Cancel' a text should NOT appear under 'Evidence file/URL'.
6. Fixed 'Cancel' button on Add popup. 

BE:
1. Added Text null able field "notes" to Evidence class
2. Created migration for "notes"
# Sanity checklist

- [x] I have clicked through the app to make sure my changes work and not break the app.
- [x] I have applied the correct milestone and labels.
- [x] My changes fix the issue described in the description (and do nothing else). 🤞
- [x] My changes are covered by tests.
- [x] My changes follow our [performance guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/performance.rst).
- [x] My changes follow our [js](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/javascript.rst) and/or [python](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/python.rst) guidelines.
- [x] My commits follow our [commit guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/how_to_write_a_commit_message.rst).

<!-- If your PR includes a migration include the additional checklist items
# Migration checklist
- [ ] Migration passes all checks from our [PR review guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/git/reviewing_pull_requests.rst#reviewing-a-pr-containing-database-migration-scripts)
- [ ] Upon merging, add 'check migration chain' and 'kokoro:force-run' labels to all open PRs with a label 'migration'
- [ ] Upon merging, update 'table structure changes' in weekly deployment documentation
-->

# PR Review checklist

- [x] The changes fix the issue and don't cause any apparent regressions.
- [x] Labels and milestone are correctly set.
- [x] The solution description matches the changes in the code.
- [x] There is no apparent way to improve the performance & design of the new code.
- [x] The pull request is opened against the correct base branch.
- [ ] Upon merging, the Jira ticket's fixversion is correctly set and the ticket is moved to "QA - In Progress".

<!-- If your code is not finished yet can include a TODO check list
# TODO

- [ ] First item on the TODO
-->

natalya-mihalyonok commented:
It will be rebased when BE PR #10038  will be merged

pavelglebov commented:
Please update description to contain FE and BE parts.

NikitaKulik commented on src/ggrc/migrations/versions/20190924_394f80f98f02_add_notes_column_to_evidence.py:
```diff
@@ -0,0 +1,30 @@
+  """Upgrade database schema and/or data, creating a new revision."""
+  op.execute("""
+      ALTER TABLE evidence
+        ADD notes TEXT;
```
Please replace with "TEXT NULL"

NikitaKulik commented on src/ggrc/models/evidence.py:
```diff
@@ -59,6 +59,8 @@ class Evidence(Roleable, Relatable, mixins.Titled,
   description = deferred(db.Column(db.Text, nullable=False, default=u""),
                          "Evidence")
 
+  notes = deferred(db.Column(db.Text, nullable=True, default=u""), "Notes")
```
Please remove default='', because in that case we have several empty parameters NULL and ""

Klim711 commented on src/ggrc-client/js/components/assessment/info-pane/create-url.js:
```diff
@@ -51,5 +51,8 @@ export default canComponent.extend({
           });
         });
     },
+    clear: function () {
```
Can be just `clear() {`

natalya-mihalyonok replied:
I've just followed style of the file. I will fix if it's required

natalya-mihalyonok commented:
@Klim711 FE part was updated according to your comments.
1. Updated styles based on mock-up. The width of new input is the same as for Add Url input. 
2. Added `word-break` for long text.
3. Fixed jumping layout when long text.

Klim711 commented on src/ggrc-client/styles/components/mapped-objects/_mapped-objects.scss:
```diff
@@ -149,6 +149,28 @@
+.item-edit-control {
+  width: max-content;
+  &__content {
+    padding: 8px 4px;
```
Left, right and bottom padding still doesn't correspond to mockups.

<img width="1129" alt="Screen Shot 2019-10-04 at 18 34 16" src="https://user-images.githubusercontent.com/10946494/66220394-9a243f00-e6d5-11e9-963d-c21a29aff792.png">

natalya-mihalyonok replied:
fixed

Klim711 commented on src/ggrc-client/styles/components/mapped-objects/_mapped-objects.scss:
```diff
@@ -149,6 +149,28 @@
+    margin: 0;
+    width: max-content;
+    &__link {
+      max-width: 210px;
```
Why link have shorter width than notes?

natalya-mihalyonok replied:
fixed

Klim711 commented on src/ggrc-client/js/components/item-edit-control/item-edit-control.stache:
```diff
@@ -0,0 +1,50 @@
+   {{#unless isEditIconDenied}}
+    <a on:el:click="setEditMode(scope.event)">
+      <action-toolbar-control>
+        <i class="fa fa-pencil-square-o"></i>
```
Icon should not have underline effect on hover.

<img width="1133" alt="Screen Shot 2019-10-04 at 18 33 02" src="https://user-images.githubusercontent.com/10946494/66221131-081d3600-e6d7-11e9-8a94-f19b66337800.png">

natalya-mihalyonok replied:
fixed

natalya-mihalyonok commented:
@Klim711 The issues were fixed, please review this again

Klim711 commented on src/ggrc-client/styles/components/mapped-objects/_mapped-objects.scss:
```diff
@@ -149,6 +149,28 @@
+    margin: 0;
+    width: max-content;
+    &__link {
+      max-width: 220px;
```
Is there a reason to set `max-width: 220px`? Let's use `max-width: 100%`.

Klim711 replied:
If restrict width of wrapper then it will be ok.

Here is my suggestion:
```
.item-edit-control {
  margin-bottom: 16px;
  width: 60%;

  &__content {
    padding: 8px 10px;
    background-color: $white;
    border-radius: 2px 2px 0 2px;
    border: 1px solid #ccc;
    margin: 0;

    &__link {
      max-width: 100%;
      font-style: italic;
      font-size: 13px;
      display: inline-block;
      cursor: pointer;

      @include text-overflow();
    }

    input {
      box-sizing: border-box;
      height: 28px;
      margin: 8px 0 0;
      width: 100%;
    }
  }
}
```

But it will break styles for url create-form, if unify template for edit and adding of urls it will be ok.

Klim711 commented on src/ggrc-client/js/components/related-objects/templates/related-urls.stache:
```diff
@@ -23,27 +46,32 @@
-          </div>
+            <div class="item-edit-control">
+              <div class="item-edit-control__content">
+                <input el:value:bind="value"
```
There is still inaccuracy with bottom padding. Let's nullify bottom margin for input (currently it is 9px) and add bottom padding in 8px for its wrapper.

natalya-mihalyonok replied:
fixed

Klim711 replied:
But there should be space between url and input.

Klim711 commented on src/ggrc-client/js/components/item-edit-control/item-edit-control.stache:
```diff
@@ -0,0 +1,46 @@
+      </div>
+    </div>
+  </div>
+  <div class="action-toolbar__controls">
```
Is it ok that action-toolbar__controls (edit, delete icons) is aligned by center of text, shouldn't it be at the top?

<img width="1040" alt="Screen Shot 2019-10-09 at 12 14 00" src="https://user-images.githubusercontent.com/10946494/66468267-89d1e280-ea8e-11e9-8d4a-357c0dec6c15.png">

natalya-mihalyonok replied:
I will clarify this with BA

natalya-mihalyonok replied:
Discussed: should be on the top.

Klim711 commented on src/ggrc-client/js/templates/evidence/modal_content.stache:
```diff
@@ -63,6 +63,22 @@
+          Notes
+        </label>
+        <div class="input-wrapper">
+          <input type="text"
```
Let's make full available width.

<img width="1669" alt="Screen Shot 2019-10-09 at 12 18 23" src="https://user-images.githubusercontent.com/10946494/66468664-34e29c00-ea8f-11e9-8a36-eff8eda1bece.png">

natalya-mihalyonok replied:
fixed

Klim711 commented on src/ggrc-client/js/templates/evidence/info.stache:
```diff
@@ -44,6 +44,16 @@
               </div>
+              <div class="row-fluid wrap-row">
+                <div class="span4">
+                  <div class="info-pane__section-title">Notes</div>
```
Let's make full available width.

Should it be placed after description?

<img width="1174" alt="Screen Shot 2019-10-09 at 12 22 48" src="https://user-images.githubusercontent.com/10946494/66469027-c6eaa480-ea8f-11e9-88cb-13db02437c00.png">

Klim711 replied:
Let's add `word-break: break-word;` for notes.

<img width="1549" alt="Screen Shot 2019-10-09 at 12 26 14" src="https://user-images.githubusercontent.com/10946494/66469199-1f21a680-ea90-11e9-8620-b90157e00155.png">

natalya-mihalyonok replied:
@Klim711 there were fixed

NikitaKulik commented:
@victoria-kostyukova 
Please check selenium tests

victoria-kostyukova commented:
@SergeyRabtsau please review the commit with selenium tests fix https://github.com/google/ggrc-core/pull/10052/commits/a6163b6b704d93c6458c1b9962171adafe1434cf

SergeyRabtsau commented:
:+1: Selenium part LGTM :tada:
